Spigot-Server-Patches/0325-World-EntityHuman-Lookup-Optimizations.patch

```diff
@@ -0,0 +1,96 @@
+     public final List<EntityHuman> players = Lists.newArrayList();
++    // Paper start - World EntityHuman Lookup Optimizations
++    public final HashMap<UUID, Entity> entitiesByUUID = com.google.common.collect.Maps.newHashMap();
++    public final HashMap<String, Entity> entitiesByName = com.google.common.collect.Maps.newHashMap();
++    // Paper end
+     public final List<Entity> k = Lists.newArrayList();
+     protected final IntHashMap<Entity> entitiesById = new IntHashMap();
+     private final long G = 16777215L;
+@@ -1060,6 +1064,10 @@ public abstract class World implements GeneratorAccess, IIBlockAccess, AutoClose
+                 EntityHuman entityhuman = (EntityHuman) entity;
+ 
+                 this.players.add(entityhuman);
++                // Paper start - World EntityHuman Lookup Optimizations
++                this.entitiesByUUID.put(entityhuman.getUniqueID(), entityhuman);
++                this.entitiesByName.put(entityhuman.getName(), entityhuman);
++                // Paper end
+                 this.everyoneSleeping();
+             }
+ 
+@@ -2659,6 +2667,8 @@ public abstract class World implements GeneratorAccess, IIBlockAccess, AutoClose
+ 
+     @Nullable
+     public EntityHuman a(String s) {
++        // Paper start - World EntityHuman Lookup Optimizations
++        /*
+         for (int i = 0; i < this.players.size(); ++i) {
+             EntityHuman entityhuman = (EntityHuman) this.players.get(i);
+ 
+@@ -2668,10 +2678,16 @@ public abstract class World implements GeneratorAccess, IIBlockAccess, AutoClose
+         }
+ 
+         return null;
++        */
++        Entity entity = this.entitiesByName.get(s);
++        return entity instanceof EntityHuman ? (EntityHuman) entity : null;
++        // Paper end
+     }
+ 
+     @Nullable
+     public EntityHuman b(UUID uuid) {
++        // Paper start - World EntityHuman Lookup Optimizations
++        /*
+         for (int i = 0; i < this.players.size(); ++i) {
+             EntityHuman entityhuman = (EntityHuman) this.players.get(i);
+ 
+@@ -2681,6 +2697,10 @@ public abstract class World implements GeneratorAccess, IIBlockAccess, AutoClose
+         }
+ 
+         return null;
++        */
++        Entity entity = this.entitiesByUUID.get(uuid);
++        return entity instanceof EntityHuman ? (EntityHuman) entity : null;
++        // Paper end
+     }
+ 
+     public void checkSession() throws ExceptionWorldConflict {
+diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
+index b57e1ff36..e6cda861f 100644
+--- a/src/main/java/net/minecraft/server/WorldServer.java
++++ b/src/main/java/net/minecraft/server/WorldServer.java
+@@ -1024,6 +1024,8 @@ public class WorldServer extends World implements IAsyncTaskHandler {
+         super.b(entity);
+         this.entitiesById.a(entity.getId(), entity);
+         // Paper start
++        this.entitiesByUUID.put(entity.getUniqueID(), entity);
```

Comment:
Oh I see it is here too, well uuid is already handled, doesn't need to be here. 

Only need name, and only for EntityHuman references, and then handle removals.
So you can filter entitiesByName to EntityHuman because we dont have profile names on non humans